packages/analytics/shippers/gainsight/src/gainsight_shipper.ts

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type {
+  AnalyticsClientInitContext,
+  EventContext,
+  Event,
+  IShipper,
+} from '@kbn/analytics-client';
+import type { GainsightApi } from './types';
+import type { GainsightSnippetConfig } from './load_snippet';
+import { loadSnippet } from './load_snippet';
+
+/**
+ * gainsight shipper.
+ */
+export class GainsightShipper implements IShipper {
+  /** Shipper's unique name */
+  public static shipperName = 'Gainsight';
+  private lastClusterName: string | undefined;
+  private readonly gainsightApi: GainsightApi;
+
+  /**
+   * Creates a new instance of the gainsightShipper.
+   * @param config {@link GainsightSnippetConfig}
+   * @param initContext {@link AnalyticsClientInitContext}
+   */
+  constructor(
+    config: GainsightSnippetConfig,
+    private readonly initContext: AnalyticsClientInitContext
+  ) {
+    const { ...snippetConfig } = config;
+    this.gainsightApi = loadSnippet(snippetConfig);
+  }
+
+  /**
+   * Calls track or set on the fields provided in the newContext.
+   * @param newContext The full new context to set {@link EventContext}
+   */
+  public extendContext(newContext: EventContext): void {
+    this.initContext.logger.debug(`Received context ${JSON.stringify(newContext)}`);
+
+    // gainsight requires different APIs for different type of contexts.
+    const { userId, cluster_name: clusterName } = newContext;
+
+    if (clusterName && clusterName !== this.lastClusterName) {
+      this.initContext.logger.debug(`Calling identify with userId ${userId}`);
+      // We need to call the API for every new userId (restarting the session).
+      this.gainsightApi('identify', {
+        id: clusterName,
+        userType: 'deployment',
+      });
+      this.gainsightApi('set', 'globalContext', {
+        kibanaUserId: userId,
+      });
```

Comment:
I think we can set this outside the if, and always update the `kibanaUserId` context.

---

Do we want to provide more data in the globalContext? We currently have info like the cluster UUID, Kibana version, stack license (might not be relevant on Cloud), Cloud ID, whether the deployment is owned by an Elastician, if the deployment is running on a Cloud trial license, the current Kibana application, and much more.